### DISCUSSION TITLE
Are all installed plugins intended to be enabled by default?

### DISCUSSION THREAD

**Discussion by Kodiologist** [pytest-dev/pytest#9928] [Q&A] [answered]
I'm not very familiar with plugins, but I installed the plugin `pytest-asyncio` for one project. Now it seems to run whenever I run pytest for any project, whether or not I add configuration for it. Is this a bug in `pytest-asyncio`, or is it a bug in `pytest`, or did I misconfigure something, or am I expected to use a virtualenv for every project if I don't want every installed plugin to run for every project? I should think that a plugin's use should be opt-in rather than automatic by virtue of being installed. Automatically enabling plugins seems like trouble waiting to happen.

**Comment by nicoddemus** [comment, marked as answer]
Hi @Kodiologist 

That's how plugins work, they are enabled by default. Some plugins are enabled depending if some other configuration takes place, like using a specific mark or flag.

**Reply by Kodiologist** [reply]
Perhaps the plugin system should provide a general way to enable and disable plugins, with all plugins disabled by default. That seems less surprising and less dangerous.

**Reply by The-Compiler** [reply]
It does already have a way to disable plugins (`-p no:pluginname`). Note that it's usually a good idea to use different [virtual environments](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/) for different projects (for a variety of reasons, not just this), so I'd argue you should typically not install pytest plugins system-wide if you don't want them to be used for every project you work on.

**Comment by nicoddemus** [comment]
> Perhaps the plugin system should provide a general way to enable and disable plugins

You can pass `-p no:<pluginname>` to disable them

> with all plugins disabled by default

Not sure, this is a backwards incompatible change that would break probably every test suite out there.

**Reply by RonnyPfannschmidt** [reply]
It would be nice to deprecate the pytest11 entrypoint and to have opt in and autoload plugins

**Reply by nicoddemus** [reply]
Not sure it is a good idea to deprecate this, as it would basically break every plugin.

**Reply by The-Compiler** [reply]
Do we really need a new entry point? All that would be needed would be some kind of `autoload_plugins = False` setting for testsuites, which would then only load plugins listed under `required_plugins`, right?

**Reply by RonnyPfannschmidt** [reply]
well, strictly speaking i would only warn for plugins requiring pytest >=8.0
having nice and separate entry points for opt in, opt out would be nice, same goes for having a more beautiful entrypint name (this is purely cosmetic)

even more nice would be to have the dependencies of the package under test determine the plugins to enable for "conservative" autoloading
